google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -822,12 +830,44 @@ private void put(SpannerException e) {
     }
 
     private PooledSession take() throws SpannerException {
-      SessionOrError s = Uninterruptibles.takeUninterruptibly(waiter);
+      SessionOrError s = takeUninterruptiblyWithTimeout();
```

Comment:
how come this needs to change?

Reply:
The `Uninterruptibles.takeUninterruptibly(waiter)` will block indefinitely if the resource pool is exhausted because of a session leak. In order to be able to log that a session request is blocked, we need a trigger that will end the trace span and do the actual logging. That is what this timeout does.

This timeout is currently also propagated to the user as a DEADLINE_EXCEEDED `SpannerException`. We could choose to handle this internally and retry the session request after logging the timeout. That would retain the current behavior where the application will potentially block indefinitely in case of a session leak. My opinion is however that it would be better to propagate this error to the user in order to give the user more information on what is going on.